### PR TITLE
Project structure and c++ trace file reader/writer

### LINKED ISSUE #1: Add a c++ trace file reader/writer
As discussed this repo should contain a writer/reader etc. for trace files (including .mcap). Here are some loose requirements I put together based and previous discussions. Please feel free to challenge something or request to add your requirements.  

## Requirements (WIP)

- **File Format Support:**
  - Should support .osi, .txth, and the upcoming .mcap, aiming for transparent integration for the user—if feasible.
  - Note that only .mcap allows index-based operations, multiple top-level messages, etc., so careful consideration is needed for handling these differences.

- **Future Python Compatibility:**
  - Design the interface with the possibility of supporting a Python version in the future.

- **Basic CI/Test Setup:**
  - Implement a minimal Continuous Integration and test setup for the repository.

- **Basic Documentation/Guide:**
  - Create an initial tutorial or usage guide. This should cover basic usage and setup.
  - More comprehensive documentation should be addressed in a separate issue to avoid overloading the PR and discussions.

- **Library and Examples:**
  - Provide the functionality as a library along with a few practical examples.

- **OSI Integration:**
  - If time permits, explore a user-friendly way to integrate the OSI (main repository) both statically and dynamically, considering time constraints for the initial setup. For a PoC/WiP it might be enough to link statically

### PR DESCRIPTION
**Pull request by TimmRuppert**
**Description**
- Adds the general project structure (not meant to be for all eternity) 
- Adds c++ tracefile writer/reader with support for different formats 

I left out further documentation/tutorials besides doxygen documentation and multiple example applications. The PR already contains a lot more than "just adding a reader/writer". I recommend that decisions about how to host, design, or set up additional tutorials and documentation be made outside of this PR. This also applies to contribution guide lines etc.

**What is missing to get out of Draft**
- [x] Provide .txth reader
- [x] Provide .txth writer
- [x] Provide .osi writer
- [x] Provide .mcap reader
- [x] Add documentation for all classes
- [x] Add more unit tests for recently added readers and writers
- [x] Unify behavior when reading past file end
- [x] Add more sanity checks (opening a opend file etc.)
- [x] Incorporate latest osi mcap spec changes concerning metadata of mcap
- [x] Incorporate tracefile naming convention in examples
- [x] Expose certain elements via API (e.g. Adding more metadata to mcap)
- [x] Solve things marked with "todo"
- [x] Cleanup interface for functions not present in all inherited classes
- [x] Fix warnings
- [x] Rename include guard to follow styleguide
- [x] Fix readme: URL osi doc, features and protobuff installation, mention examples

Closes #1

### CODE REVIEW

File: include/osi-utilities/tracefile/reader/txthTraceFileReader.h
Comment (ClemensLinnhoff) [inline include/osi-utilities/tracefile/reader/txthTraceFileReader.h:1]: Capitalize filename to be consistent with the others (also for the corresponding writer and cpp files.

**Comment by ClemensLinnhoff** [comment]
I generally tested all examples and they all worked with my test files.

File: include/osi-utilities/tracefile/reader/txthTraceFileReader.h
Comment (TimmRuppert) [reply include/osi-utilities/tracefile/reader/txthTraceFileReader.h:1]: Done. 

BTW. I intentionally named the .osi files reader `NativeBinaryTraceFileReader` instead of `NativeBinaryTraceFileReader` to avoid giving the impression that it’s a general trace file reader for the OSI interface, rather than one specifically for .osi files. 

So that one is not really consistent with the others but IMHO makes more sense that way.

**Comment by TimmRuppert** [comment]
> Just a question that come to mind without research. Did we take into account that other signals like ROS messages etc can be placed alongside the osi traces?

Yes. 

For the reader: 
- You have to manually set an [option](https://github.com/OpenSimulationInterface/osi-utilities/blob/794c67cd324360b874e7b5182d0c1813cc87d2b8/include/osi-utilities/tracefile/reader/MCAPTraceFileReader.h#L57) to skip those, otherwise you will get an error while calling the read() function to get the next osi message.

For the writer:
- The underlying mcap writer can be [retrieved](https://github.com/OpenSimulationInterface/osi-utilities/blob/794c67cd324360b874e7b5182d0c1813cc87d2b8/include/osi-utilities/tracefile/writer/MCAPTraceFileWriter.h#L85) and thus you can do to it whatever mcap allows (e.g. adding non-osi messages) 

The [unittest](https://github.com/OpenSimulationInterface/osi-utilities/blob/794c67cd324360b874e7b5182d0c1813cc87d2b8/tests/src/tracefile/reader/MCAPTraceFileReader_test.cpp#L59) of the reader creates an mcap with protobuf and json messages for testing.

**Review by jdsika — APPROVED** [review]
Fix metadata in converter

**Comment by jdsika** [comment]
Please review this PR to be compliant with current **final** [draft](https://github.com/OpenSimulationInterface/open-simulation-interface/pull/841/files)

**Comment by ClemensLinnhoff** [comment]
Since the zero_time found its way into the OSI MCAP specification as mandatory metadata, this needs to be implemented in osi-utils.
@jdsika, @pmai where shall this information come from? Can we extract it from an OSI field contained in the data? Or does the user have to set it manually, and if so, is there a default value, if the user does not set it?

**Comment by jdsika** [comment]
On a virgin Ubuntu:

`asam-osi-utilities/include/osi-utilities/tracefile/reader/MCAPTraceFileReader.h:9:10: fatal error: mcap/reader.hpp: No such file or directory`

**Comment by jdsika** [comment]
How do you decide in how many chunks the file will be separated? The osi_to_mcap does work in general:

image

**Comment by TimmRuppert** [comment]
> How do you decide in how many chunks the file will be separated? 


The default upstream mcap [chunk size](https://github.com/foxglove/mcap/blob/bebea860f68b278d6cccdb70e0ed299d2656af96/cpp/mcap/include/mcap/types.hpp#L29) is 1024 * 768

You can manually set the mcap options including chunksize, compression etc. 

```c++
    auto trace_file_writer = osi3::MCAPTraceFileWriter();
    const auto trace_file_path = "/tmp/my_different_chunk_size.mcap"

    mcap::McapWriterOptions options("test");
    options.chunkSize = 10000;
    options.compression = mcap::Compression::Lz4;

    trace_file_writer.Open(trace_file_path, options);

```

File: examples/convert_osi2mcap.cpp [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        return 1;
+    }
+
+    auto trace_file_writer = osi3::MCAPTraceFileWriter();
+    if (!trace_file_writer.Open(options->output_file_path)) {
+        std::cerr << "ERROR: Could not open output file " << options->output_file_path << std::endl;
+        return 1;
+    }
+
+    // according to the OSI specification mcap must contain a timestamp and zero_time metadata entry
```
Comment (ClemensLinnhoff) [inline examples/convert_osi2mcap.cpp:172]: This is not true anymore. zero_time is no longer mandatory. Trying to parse it from the filename is still a good idea.

File: examples/convert_osi2mcap.cpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    // try to parse the file timestamp from the .osi file name (as it should follow the recommended OSI naming conventions)
```
Comment (ClemensLinnhoff) [inline examples/convert_osi2mcap.cpp:173]: What happens if the naming convention is not used in the .osi trace file?

Comment (ClemensLinnhoff) [reply examples/convert_osi2mcap.cpp:173]: Or do we require this here, because we need the message type anyways?

Comment (TimmRuppert) [reply examples/convert_osi2mcap.cpp]: See https://github.com/Lichtblick-Suite/asam-osi-utilities/pull/2#discussion_r1865433822

Now the `zero_time` is simply not set if not present in the filename.

```
$ ./convert_osi2mcap /tmp/test.osi /tmp/test_converted.mcap --input-type SensorView
Input file: "/tmp/test.osi"
Output file: "/tmp/test_converted.mcap"
Finished single channel binary to mcap converter

$ mcap list metadata /tmp/test_converted.mcap
name              	offset	length	metadata                                                                                                                                                                                                                                  	
net.asam.osi.trace	46    	282   	{"creation_time":"2024-12-02T08:37:58.1Z","description":"Converted from /tmp/test.osi","max_osi_version":"3.7.0","max_protobuf_version":"3.21.12","min_osi_version":"3.7.0","min_protobuf_version":"3.21.12","version":"3.7.0"}	
```

Comment (jdsika) [reply examples/convert_osi2mcap.cpp]: Proposal: Provide type e.g. "sv" by parameter. Throw error if not present in name and no parameter is set. Throw warning if name indicates other type then parameter. parameter overwrites name type.

Comment (TimmRuppert) [reply examples/convert_osi2mcap.cpp]: > Proposal: Provide type e.g. "sv" by parameter. Throw error if not present in name and no parameter is set. Throw warning if name indicates other type then parameter. parameter overwrites name type.

That is already implemented 

```bash
$ ./convert_osi2mcap --help
Usage: convert_osi2mcap <input_file> <output_file> [--input-type <message_type>]

Arguments:
  input_file              Path to the input OSI trace file
  output_file             Path to the output MCAP file
  --input-type <message_type>   Optional: Specify input message type if not stated in filename

Valid message types:
  MotionRequest
  TrafficCommandUpdate
  TrafficCommand
  HostVehicleData
  SensorView
  TrafficUpdate
  SensorData
  StreamingUpdate
  GroundTruth
  
$ ./convert_osi2mcap /tmp/test.osi /tmp/test_converted.mcap
Input file: "/tmp/test.osi"
Output file: "/tmp/test_converted.mcap"
ERROR: Unable to determine message type from the filename '"/tmp/test.osi"'. Please ensure the filename follows the recommended OSI naming conventions as specified in the documentation or specify the message type manually.
ERROR: Could not open input file "/tmp/test.osi

```

Comment (TimmRuppert) [reply examples/convert_osi2mcap.cpp]: > Throw warning if name indicates other type then parameter. parameter overwrites name type.

Done in https://github.com/Lichtblick-Suite/asam-osi-utilities/pull/2/commits/52215386d69ea4d186e2497ac52fff79fdede3b5 and https://github.com/Lichtblick-Suite/asam-osi-utilities/pull/2/commits/0df13a4432946a5a08f13f386e62d56c5f642e0f

**Review by ClemensLinnhoff — APPROVED** [review]

**Comment by jdsika** [comment]
Where can I see which MCAP repo tag/version is pinned for the submodule?

File: examples/example_mcap_writer.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    const auto trace_file_path = GenerateTempFilePath();
+    std::cout << "Creating trace_file at " << trace_file_path << std::endl;
+    trace_file_writer.Open(trace_file_path);
+
+    // add required and optional metadata to the net.asam.osi.trace metadata record
+    auto net_asam_osi_trace_metadata = osi3::MCAPTraceFileWriter::PrepareRequiredFileMetadata();
+    // Add optional metadata to the net.asam.osi.trace metadata record, as recommended by the OSI specification.
+    net_asam_osi_trace_metadata.metadata["description"] = "Example mcap trace file created with the ASAM OSI utilities library."; // optional field
+    net_asam_osi_trace_metadata.metadata["creation_time"] = osi3::MCAPTraceFileWriter::GetCurrentTimeAsString(); // optional field
+    net_asam_osi_trace_metadata.metadata["authors"] = "Jane Doe, John Doe"; // optional field
```
Comment (jdsika) [inline examples/example_mcap_writer.cpp:31]: Where is the rest of the mandatory metadata written/defined?

Comment (jdsika) [reply examples/example_mcap_writer.cpp:31]: Does the example fail? src/tracefile/writer/MCAPTraceFileWriter.cpp line 119 requires it?

```
        constexpr std::array<const char*, 5> kRequiredFields = {"version", "min_osi_version", "max_osi_version", "min_protobuf_version", "max_protobuf_version"};
        for (const auto& field : kRequiredFields) {
            if (metadata.metadata.find(field) == metadata.metadata.end()) {
                std::cerr << "ERROR: cannot add net.asam.osi.trac meta-data record without a " << field << " field.\n";
                return false;
            }
        }
```

**Comment by TimmRuppert** [comment]
> Where can I see which MCAP repo tag/version is pinned for the submodule?

To my knowledge submodules always point to a commit. If you checkout the tag (like I did) it will be in detached HEAD mode pointing to the commit the tag points to. 

Locally you can still see the tag 
```bash
$ cd asam-osi-utilities/lib/mcap
asam-osi-utilities/lib/mcap$ git status
HEAD detached at releases/cpp/v1.4.1
nothing to commit, working tree clean

```

File: examples/example_mcap_writer.cpp (hunk 3 above)
Comment (TimmRuppert) [reply examples/example_mcap_writer.cpp:31]: [Here](https://github.com/Lichtblick-Suite/asam-osi-utilities/pull/2/files/0df13a4432946a5a08f13f386e62d56c5f642e0f#diff-56a24a74e9814d13490bf27c09db2a08b133b14142bc35f935a798e37db98a5cR27) it gets created  and [here](https://github.com/Lichtblick-Suite/asam-osi-utilities/pull/2/files/0df13a4432946a5a08f13f386e62d56c5f642e0f#diff-56a24a74e9814d13490bf27c09db2a08b133b14142bc35f935a798e37db98a5cR32) the mandatory and optional metadata are written. 

The convenience function `osi3::MCAPTraceFileWriter::PrepareRequiredFileMetadata();` provides the required (by the current state of the new OSI spec) metadata. `PrepareRequiredFileMetadata()`  is obviously not useful for merging mcap files etc. as the min and max values of osi/protobuf version metadata will all get the same value (therefore a convenience function and not part of setting any metadata automatically).

Comment (TimmRuppert) [reply examples/example_mcap_writer.cpp:31]: > Does the example fail? src/tracefile/writer/MCAPTraceFileWriter.cpp line 119 requires it?

No because it uses `osi3::MCAPTraceFileWriter::PrepareRequiredFileMetadata()`

```bash

$ git rev-parse HEAD
0df13a4432946a5a08f13f386e62d56c5f642e0f
$ git status
On branch 1-add-a-c++-trace-file-readerwriter
Your branch is up to date with 'origin/1-add-a-c++-trace-file-readerwriter'.

nothing to commit, working tree clean

$ ./cmake-build-debug/examples/example_mcap_writer 
Starting MCAP Writer example:
Creating trace_file at "/tmp/sv_example.mcap"
Finished MCAP Writer example
```

Comment (jdsika) [reply examples/example_mcap_writer.cpp:31]: ah ok, but then this makes no sense `metadata.metadata["version"] = osi_version_string; `because it is the version of the spec (we don't have yet but we would anticipate 3.8.0 which should include the multi tracefile spec.

Comment (jdsika) [reply examples/example_mcap_writer.cpp:31]: This should (after merge and release) be the osi version which is mentioned in the CMAKE

Comment (TimmRuppert) [reply examples/example_mcap_writer.cpp:31]: > ah ok, but then this makes no sense metadata.metadata["version"] = osi_version_string; because it is the version of the spec (we don't have yet but we would anticipate 3.8.0 which should include the multi tracefile spec.

Ah you mean this should be hard-coded? Yes this makes sense. 

> ah ok, but then this makes no sense metadata.metadata["version"] = osi_version_string; because it is the version of the spec (we don't have yet but we would anticipate 3.8.0 which should include the multi tracefile spec.


In the cmake you can specify what ever osi you want but that is not related with the spec we are trying to fulfill here right? Updating the osi you build against does not automatically fix potential osi mcap schema issue which need to be fixed in the library itself.  

> we would anticipate 3.8.0

The assigned milestone is called 3.7.1

> This should (after merge and release)

Which merge and release? OSI muli trace file spec or this PR of asam-osi-utilities? I can imagine that asam-osi-utilities releases **could** be made quite complicated: multiple asam-osi-utilities versions for multiple osi versions. It might make sense to simply target the newest OSI version and always provide independent improvements and fixes - but that is a discussion for another day.  



**To sum it up**: If you do not disagree I would simply hardcode this to `3.7.0-a63f47d7b21bb44edafca176700d9cedf6b12d81` which represents the current values of Pieres last commit (the `VERSION` file in that branch + commit).

**Review by jdsika — APPROVED** [review]
Just add the last changes we discussed via teams and it is good to merge

**Comment by TimmRuppert** [comment]
> Just add the last changes we discussed via teams and it is good to merge

I added more options to the converter as requested 
```bash
  --chunk_size <size>           Optional: Chunk size in bytes (default: 786432)
  --compression <type>          Optional: Compression type (none, lz4, zstd) (default: zstd)
  --compression_level <type>    Optional: Compression level (fastest, fast, default, slow, slowest) (default: default)

```

I also did some quick tests:
- all lz4
-  the number (2*) in the filename is the chunksize
-  "default" or "fastest" in the filename  means the compression level

```bash
-rw-rw-r-- 1 timm timm 61M Dec  2 15:07 /tmp/test_200000000_lz4_default.mcap
-rw-rw-r-- 1 timm timm 80M Dec  2 15:06 /tmp/test_200000000_lz4_fastest.mcap
-rw-rw-r-- 1 timm timm 61M Dec  2 15:07 /tmp/test_20000000_lz4_default.mcap
-rw-rw-r-- 1 timm timm 80M Dec  2 15:05 /tmp/test_20000000_lz4_fastest.mcap
-rw-rw-r-- 1 timm timm 80M Dec  2 15:05 /tmp/test_2000000_lz4_fastest.mcap
-rw-rw-r-- 1 timm timm 80M Dec  2 15:04 /tmp/test_200000_lz4_fastest.mcap
-rw-rw-r-- 1 timm timm 80M Dec  2 15:04 /tmp/test_20000_lz4_fastest.mcap
-rw-rw-r-- 1 timm timm 80M Dec  2 15:04 /tmp/test_2000_lz4_fastest.mcap
-rw-rw-r-- 1 timm timm 61M Dec  2 15:04 /tmp/test_200_lz4_default.mcap
-rw-rw-r-- 1 timm timm 80M Dec  2 15:04 /tmp/test_200_lz4_fastest.mcap
-rw-rw-r-- 1 timm timm 80M Dec  2 15:03 /tmp/test_20_lz4_fastest.mcap
-rw-rw-r-- 1 timm timm 61M Dec  2 15:20 /tmp/test_2_lz4_default.mcap
-rw-rw-r-- 1 timm timm 80M Dec  2 15:03 /tmp/test_2_lz4_fastest.mcap

```

So the chunk size does not seem to have an effect :man_shrugging: 

but it is applied. Check the following 2 vs. 11 vs. 66 chunks. The file had 66 messages.

```bash
$mcap info /tmp/test_200000000_lz4_fastest.mcap
library:   libmcap 1.4.1  
profile:   osi2mcap       
messages:  66             
duration:  6.499999999s   
start:     0.100000000    
end:       6.599999999    
compression:
	lz4: [2/2 chunks] [209.92 MiB/79.75 MiB (62.01%)] [12.27 MiB/sec] 
channels:
	(1) ConvertedTrace  66 msgs (10.15 Hz)   : osi3.SensorData [protobuf]  
channels: 1
attachments: 0
metadata: 1


$ mcap info /tmp/test_20000000_lz4_fastest.mcap
library:   libmcap 1.4.1  
profile:   osi2mcap       
messages:  66             
duration:  6.499999999s   
start:     0.100000000    
end:       6.599999999    
compression:
	lz4: [11/11 chunks] [209.92 MiB/79.75 MiB (62.01%)] [12.27 MiB/sec] 
channels:
	(1) ConvertedTrace  66 msgs (10.15 Hz)   : osi3.SensorData [protobuf]  
channels: 1
attachments: 0
metadata: 1

$ mcap info /tmp/test_2000000_lz4_fastest.mcap
library:   libmcap 1.4.1  
profile:   osi2mcap       
messages:  66             
duration:  6.499999999s   
start:     0.100000000    
end:       6.599999999    
compression:
	lz4: [66/66 chunks] [209.92 MiB/79.75 MiB (62.01%)] [12.27 MiB/sec] 
channels:
	(1) ConvertedTrace  66 msgs (10.15 Hz)   : osi3.SensorData [protobuf]  
channels: 1
attachments: 0
metadata: 1

```
